lib/AST/ProtocolConformance.cpp

```diff
@@ -1059,3 +1059,151 @@ DeclContext::getLocalConformances(
+bool ProtocolConformance::isCanonical() const {
+  // Normal conformances are considered canonical by
+  // construction.
+  if (getKind() != ProtocolConformanceKind::Normal) {
+    if (!getType()->isCanonical())
+      return false;
+    if (!getInterfaceType()->isCanonical())
+      return false;
+  }
+
+  switch (getKind()) {
+  case ProtocolConformanceKind::Normal: {
+    auto normalConformance = cast<NormalProtocolConformance>(this);
+    auto sigConformances = normalConformance->getSignatureConformances();
+    for (auto conf : sigConformances)
+      if (!conf.isCanonical())
+        return false;
+    return true;
+  }
+  case ProtocolConformanceKind::Inherited: {
+    // Substitute the base.
+    auto inheritedConformance
+      = cast<InheritedProtocolConformance>(this);
+    return inheritedConformance->getInheritedConformance()->isCanonical();
+  }
+  case ProtocolConformanceKind::Specialized: {
+    // Substitute the substitutions in the specialized conformance.
+    auto spec = cast<SpecializedProtocolConformance>(this);
+    auto genericConformance = spec->getGenericConformance();
+    if (!genericConformance->isCanonical())
+      return false;
+    auto specSubs = spec->getGenericSubstitutions();
+    for (auto sub : specSubs) {
+      if (!sub.getReplacement()->isCanonical())
+        return false;
+      for (auto conf : sub.getConformances()) {
+        if (!conf.isCanonical())
+          return false;
+      }
+    }
+    return true;
+  }
+  }
+  llvm_unreachable("bad ProtocolConformanceKind");
+}
+
+Substitution Substitution::getCanonicalSubstitution(bool *wasCanonical) const {
+  bool createdNewCanonicalConformances = false;
+  bool createdCanReplacement = false;
+  SmallVector<ProtocolConformanceRef, 4> newCanConformances;
+
+  CanType canReplacement = getReplacement()->getCanonicalType();
+
+  if (!getReplacement()->isCanonical()) {
+    createdCanReplacement = true;
+  }
+
+  for (auto conf : getConformances()) {
+    if (conf.isCanonical()) {
+      newCanConformances.push_back(conf);
+      continue;
+    }
+    newCanConformances.push_back(conf.getCanonicalConformanceRef());
+    createdNewCanonicalConformances = true;
+  }
+
+  ArrayRef<ProtocolConformanceRef> canConformances = getConformances();
+  if (createdNewCanonicalConformances) {
+    auto &C = canReplacement->getASTContext();
+    canConformances = C.AllocateCopy(newCanConformances);
+  }
+
+  if (createdCanReplacement || createdNewCanonicalConformances) {
+    if (wasCanonical)
+      *wasCanonical = false;
+    return Substitution(canReplacement, canConformances);
+  }
+  if (wasCanonical)
+    *wasCanonical = true;
+  return *this;
+}
+
+SubstitutionList swift::getCanonicalSubstitutionList(SubstitutionList subs) {
+  SmallVector<Substitution, 4> newCanSubstitutions;
+  bool subListWasCanonical = true;
+  for (auto &sub : subs) {
+    bool subWasCanonical = false;
+    auto canSub = sub.getCanonicalSubstitution(&subWasCanonical);
```

Comment:
How about just checking isCanonical() on the substitution first and removing the out parameter. The conformance array will have zero or one elements most cases, so checking it is not a big deal.